File: app/src/main/java/com/wilinz/yuetingmusic/player/PlayerManager.java
package com.wilinz.yuetingmusic.player;

import android.app.Service;
import android.media.session.MediaController;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.MediaControllerCompat;
import android.support.v4.media.session.MediaSessionCompat;
import android.support.v4.media.session.PlaybackStateCompat;
import android.util.Log;

import androidx.annotation.Nullable;

import com.google.android.exoplayer2.ExoPlayer;
import com.google.android.exoplayer2.MediaItem;
import com.google.android.exoplayer2.MediaMetadata;
import com.google.android.exoplayer2.PlaybackException;
import com.google.android.exoplayer2.Player;
import com.wilinz.yuetingmusic.Key;
import com.wilinz.yuetingmusic.Pref;
import com.wilinz.yuetingmusic.constant.PlayMode;
import com.wilinz.yuetingmusic.data.model.Song;
import com.wilinz.yuetingmusic.data.repository.SongRepository;
import com.wilinz.yuetingmusic.util.LogUtil;
import com.wilinz.yuetingmusic.util.MediaUtil;
import com.wilinz.yuetingmusic.util.ToastUtilKt;

import java.util.Arrays;
import java.util.List;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import kotlin.collections.CollectionsKt;

public class PlayerManager extends MyMediaSessionCallback {

    private final static String TAG = "MediaPlayerSession";
    //    private final MediaPlayer mediaPlayer = new MediaPlayer();
    ExoPlayer exoPlayer = new ExoPlayer.Builder(service).build();

    private final MediaSessionCompat mediaSession;

    private boolean isPreparedSeek = false;
    private long preparedSeekPosition = 0;

    public PlayerManager(Service service, MyAudioManager myAudioManager, MyNotificationManager myNotificationManager, MediaSessionCompat mediaSession) {
        super(service, myAudioManager, myNotificationManager);
        this.mediaSession = mediaSession;
        initPlayer(mediaSession);
    }

    private void initPlayer(MediaSessionCompat mediaSession) {
        setPlayMode(Pref.getInstance(service.getApplication()).getPlayMode());
        exoPlayer.addListener(new Player.Listener() {

            @Override
            public void onMediaItemTransition(@Nullable MediaItem mediaItem, int reason) {
                Player.Listener.super.onMediaItemTransition(mediaItem, reason);
                if (mediaItem == null || mediaItem.localConfiguration == null) return;
                Song song = CollectionsKt.firstOrNull(SongRepository.getInstance().getPlayQueue(), song1 -> song1.getUniqueId().equals(mediaItem.mediaId));
                SongRepository.getInstance().setCurrentSong(song);
                SongRepository.getInstance().saveRecentSong(song)
                        .subscribe();
            }

            @Override
            public void onIsPlayingChanged(boolean isPlaying) {
                Player.Listener.super.onIsPlayingChanged(isPlaying);
                if (isPlaying) {
                    sendMetadata();
                    updatePlaybackState(PlaybackStateCompat.STATE_PLAYING);
                }
            }

            @Override
            public void onMediaMetadataChanged(MediaMetadata mediaMetadata) {
                Player.Listener.super.onMediaMetadataChanged(mediaMetadata);
                sendMetadata();
                updatePlaybackState(PlaybackStateCompat.STATE_PLAYING);
            }

            @Override
            public void onPlayerError(PlaybackException error) {
                Player.Listener.super.onPlayerError(error);
                error.printStackTrace();
                updatePlaybackState(PlaybackStateCompat.STATE_PAUSED);
//                exoPlayer.setMediaItem(exoPlayer.getCurrentMediaItem(), exoPlayer.getCurrentPosition());
//                exoPlayer.prepare();
//                onPlay();
                Log.d(TAG, "onPlayerError: ");
            }

            @Override
            public void onPlaybackStateChanged(int playbackState) {
                Player.Listener.super.onPlaybackStateChanged(playbackState);
                updatePlaybackState();
            }

            @Override
            public void onPositionDiscontinuity(Player.PositionInfo oldPosition, Player.PositionInfo newPosition, int reason) {
                Player.Listener.super.onPositionDiscontinuity(oldPosition, newPosition, reason);
                updatePlaybackState();
            }

        });
//        mediaPlayer.setOnPreparedListener(this::onPreparedListener);
//        mediaPlayer.setOnCompletionListener(this::onCompletionListener);
    }

    private void setPlayMode(int playMode) {
        switch (playMode) {
            case PlayMode.SINGLE_LOOP:
                exoPlayer.setShuffleModeEnabled(false);
                exoPlayer.setRepeatMode(Player.REPEAT_MODE_ONE);
                break;
            case PlayMode.ORDERLY:
                exoPlayer.setShuffleModeEnabled(false);
                exoPlayer.setRepeatMode(Player.REPEAT_MODE_ALL);
                break;
            case PlayMode.SHUFFLE:
                exoPlayer.setShuffleModeEnabled(true);
                exoPlayer.setRepeatMode(Player.REPEAT_MODE_ALL);
        }
        Bundle bundle = new Bundle();
        bundle.putInt(Key.playMode, playMode);
        updatePlaybackState(bundle);
    }

    private void updatePlaybackState(Bundle... bundle) {
        updatePlaybackState(exoPlayer.isPlaying() ? PlaybackStateCompat.STATE_PLAYING : PlaybackStateCompat.STATE_PAUSED, bundle);
    }

    private void updatePlaybackState(@PlaybackStateCompat.State int state, Bundle... bundle) {
        // 更新视频的总进度, setMetadata 会更新MediaControlCompat的onMetadataChanged
        PlaybackStateCompat mPlaybackStateCompat = new PlaybackStateCompat.Builder()
                .setState(state,
                        exoPlayer.getCurrentPosition(),
                        1.0f)
                .setActions(getAvailableActions(state))
                .setExtras(CollectionsKt.getOrNull(Arrays.asList(bundle), 0))
                .build();
        mediaSession.setPlaybackState(mPlaybackStateCompat);
    }

    public static long getAvailableActions(@PlaybackStateCompat.State int state) {
        long actions = (PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS
                | PlaybackStateCompat.ACTION_SKIP_TO_NEXT
                | PlaybackStateCompat.ACTION_REWIND
                | PlaybackStateCompat.ACTION_SEEK_TO);

        if (state == PlaybackStateCompat.STATE_PLAYING) {
            actions = actions | PlaybackStateCompat.ACTION_PAUSE;
        } else {
            actions = actions | PlaybackStateCompat.ACTION_PLAY;
        }
        return actions;
    }


    @Override
    public void onPlay() {
        super.onPlay();
        if (getPlaybackState().getState() != PlaybackStateCompat.STATE_PLAYING
                && myAudioManager.requestAudioFocus()
        ) {
            register();
            exoPlayer.play();
        }
    }

    public void sendMetadata() {
        MediaItem mediaItem = exoPlayer.getCurrentMediaItem();
        MediaMetadataCompat mediaMetadataCompat = null;
        if (mediaItem == null) {
            Song currentSong = SongRepository.getInstance().getCurrentSong();
            if (currentSong != null) {
                mediaMetadataCompat = currentSong.mapToMediaMetadata(exoPlayer.getCurrentPosition(), exoPlayer.getDuration());
            }
        } else {
            mediaMetadataCompat = MediaUtil.getMediaMetadataCompat(mediaItem, exoPlayer.getContentDuration());
        }
        if (mediaMetadataCompat != null) {
            mediaSession.setMetadata(mediaMetadataCompat);
        }
    }

    private PlaybackStateCompat getPlaybackState() {
        return mediaSession.getController().getPlaybackState();
    }

    @Override
    public void onPause() {
        super.onPause();
        LogUtil.d(TAG, "onPause");
        if (getPlaybackState().getState() == PlaybackStateCompat.STATE_PLAYING) {
            exoPlayer.pause();
            updatePlaybackState(PlaybackStateCompat.STATE_PAUSED);
            myAudioManager.unregisterBecomingNoisyReceiver();
            myAudioManager.abandonAudioFocus();
        }
    }

    @Override
    public void onStop() {
        super.onStop();
        exoPlayer.stop();
    }

    @Override
    public void onPlayFromUri(Uri uri, Bundle extras) {
        super.onPlayFromUri(uri, extras);
        List<Song> songs = extras.getParcelableArrayList(Key.songList);
        if (songs != null) {
            int index = CollectionsKt.indexOfFirst(songs, song -> song.url.equals(uri.toString()));
            SongRepository.getInstance().setPlayQueue(songs);
            List<MediaItem> mediaItems = CollectionsKt.map(songs, Song::mapToExoPlayerMediaItem);
            exoPlayer.setMediaItems(mediaItems, index, 0);
        } else {
            exoPlayer.setMediaItem(MediaItem.fromUri(uri));
        }

        updatePlaybackState(PlaybackStateCompat.STATE_CONNECTING);
        exoPlayer.prepare();
        onPlay();
    }


    @Override
    public void onSkipToNext() {
        super.onSkipToNext();
        exoPlayer.seekToNext();
        sendMetadata();
    }

    @Override
    public void onSkipToPrevious() {
        super.onSkipToPrevious();
        exoPlayer.seekToPrevious();
        sendMetadata();
    }

    @Override
    public void onSeekTo(long pos) {
        super.onSeekTo(pos);
        Log.d(TAG, "onSeekTo: " + pos);
        exoPlayer.seekTo(pos);
        updatePlaybackState(PlaybackStateCompat.STATE_PLAYING);
    }

    @Override
    public void onSetPlaybackSpeed(float speed) {
        super.onSetPlaybackSpeed(speed);
    }

    @Override
    public void onSetRepeatMode(int repeatMode) {
        super.onSetRepeatMode(repeatMode);
        int playModel = repeatMode == PlaybackStateCompat.REPEAT_MODE_ONE ? PlayMode.SINGLE_LOOP : PlayMode.ORDERLY;
        setPlayMode(playModel);
        Pref.getInstance(service.getApplication()).setPlayMode(playModel);
    }

    @Override
    public void onSetShuffleMode(int shuffleMode) {
        super.onSetShuffleMode(shuffleMode);
        setPlayMode(PlayMode.SHUFFLE);
        Pref.getInstance(service.getApplication()).setPlayMode(PlayMode.SHUFFLE);
    }

    @Override
    public void unregister() {
        super.unregister();
        exoPlayer.release();
        mediaSession.release();
    }

    @SuppressWarnings("all")
    @Override
    public void onCustomAction(String action, Bundle extras) {
        super.onCustomAction(action, extras);
        switch (action) {
            case ACTION_SAVE_CURRENT_SONG_TO_FAVORITE:
                MediaItem mediaItem = exoPlayer.getCurrentMediaItem();
                if (mediaItem == null || mediaItem.localConfiguration == null) return;
                Song song = CollectionsKt.firstOrNull(SongRepository.getInstance().getPlayQueue(), song1 -> song1.getUniqueId().equals(mediaItem.mediaId));
                SongRepository.getInstance().saveFavoriteSong(SongRepository.getInstance().getCurrentSong())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe((song1) -> {
                            ToastUtilKt.toast(service, "收藏成功");
                        }, e -> {
                            e.printStackTrace();
                            ToastUtilKt.toast(service, "收藏失败：" + e.getMessage());
                        });
                break;
        }
    }

    public final static String ACTION_SAVE_CURRENT_SONG_TO_FAVORITE = "action_save_to_favorite";

    public static void switchPlayMode(MediaControllerCompat.TransportControls transportControls, int playMode) {
        if (playMode == PlayMode.ORDERLY) {
            transportControls.setRepeatMode(PlaybackStateCompat.REPEAT_MODE_ONE);
        } else if (playMode == PlayMode.SINGLE_LOOP) {
            transportControls.setShuffleMode(PlaybackStateCompat.SHUFFLE_MODE_ALL);
        } else {
            transportControls.setRepeatMode(PlaybackStateCompat.REPEAT_MODE_ALL);
        }
    }

}

File: app/src/main/java/com/wilinz/yuetingmusic/player/PlayerManager.kt
package com.wilinz.yuetingmusic.player

import android.app.Service
import android.net.Uri
import android.os.Bundle
import android.support.v4.media.MediaMetadataCompat
import android.support.v4.media.session.MediaControllerCompat
import android.support.v4.media.session.MediaSessionCompat
import android.support.v4.media.session.PlaybackStateCompat
import android.util.Log
import com.google.android.exoplayer2.*
import com.google.android.exoplayer2.Player.PositionInfo
import com.wilinz.yuetingmusic.Key
import com.wilinz.yuetingmusic.Pref.Companion.getInstance
import com.wilinz.yuetingmusic.constant.PlayMode
import com.wilinz.yuetingmusic.data.model.Song
import com.wilinz.yuetingmusic.data.repository.SongRepository.Companion.instance
import com.wilinz.yuetingmusic.util.LogUtil.d
import com.wilinz.yuetingmusic.util.MediaUtil.getMediaMetadataCompat
import com.wilinz.yuetingmusic.util.toast
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import java.util.*

class PlayerManager(
    service: Service?,
    myAudioManager: MyAudioManager?,
    myNotificationManager: MyNotificationManager?,
    private val mediaSession: MediaSessionCompat
) : MyMediaSessionCallback(
    service!!, myAudioManager!!, myNotificationManager!!
) {
    //    private final MediaPlayer mediaPlayer = new MediaPlayer();
    var exoPlayer = ExoPlayer.Builder(service!!).build()
    private val isPreparedSeek = false
    private val preparedSeekPosition: Long = 0
    private fun initPlayer(mediaSession: MediaSessionCompat) {
        setPlayMode(getInstance(service.application)!!.playMode)
        exoPlayer.addListener(object : Player.Listener {
            override fun onMediaItemTransition(mediaItem: MediaItem?, reason: Int) {
                super.onMediaItemTransition(mediaItem, reason)
                if (mediaItem == null || mediaItem.localConfiguration == null) return
                val song =
                    instance!!.playQueue.firstOrNull { song1: Song -> song1.uniqueId == mediaItem.mediaId }
                instance!!.currentSong = song
                instance!!.saveRecentSong(song!!)
                    .subscribe()
            }

            override fun onIsPlayingChanged(isPlaying: Boolean) {
                super.onIsPlayingChanged(isPlaying)
                if (isPlaying) {
                    sendMetadata()
                    updatePlaybackState(PlaybackStateCompat.STATE_PLAYING)
                }
            }

            override fun onMediaMetadataChanged(mediaMetadata: MediaMetadata) {
                super.onMediaMetadataChanged(mediaMetadata)
                sendMetadata()
                updatePlaybackState(PlaybackStateCompat.STATE_PLAYING)
            }

            override fun onPlayerError(error: PlaybackException) {
                super.onPlayerError(error)
                error.printStackTrace()
                updatePlaybackState(PlaybackStateCompat.STATE_PAUSED)
                //                exoPlayer.setMediaItem(exoPlayer.getCurrentMediaItem(), exoPlayer.getCurrentPosition());
//                exoPlayer.prepare();
//                onPlay();
                Log.d(TAG, "onPlayerError: ")
            }

            override fun onPlaybackStateChanged(playbackState: Int) {
                super.onPlaybackStateChanged(playbackState)
                updatePlaybackState()
            }

            override fun onPositionDiscontinuity(
                oldPosition: PositionInfo,
                newPosition: PositionInfo,
                reason: Int
            ) {
                super.onPositionDiscontinuity(oldPosition, newPosition, reason)
                updatePlaybackState()
            }
        })
        //        mediaPlayer.setOnPreparedListener(this::onPreparedListener);
//        mediaPlayer.setOnCompletionListener(this::onCompletionListener);
    }

    private fun setPlayMode(playMode: Int) {
        when (playMode) {
            PlayMode.SINGLE_LOOP -> {
                exoPlayer.shuffleModeEnabled = false
                exoPlayer.repeatMode = Player.REPEAT_MODE_ONE
            }
            PlayMode.ORDERLY -> {
                exoPlayer.shuffleModeEnabled = false
                exoPlayer.repeatMode = Player.REPEAT_MODE_ALL
            }
            PlayMode.SHUFFLE -> {
                exoPlayer.shuffleModeEnabled = true
                exoPlayer.repeatMode = Player.REPEAT_MODE_ALL
            }
        }
        val bundle = Bundle()
        bundle.putInt(Key.playMode, playMode)
        updatePlaybackState(bundle)
    }

    private fun updatePlaybackState(vararg bundle: Bundle) {
        updatePlaybackState(
            if (exoPlayer.isPlaying) PlaybackStateCompat.STATE_PLAYING else PlaybackStateCompat.STATE_PAUSED,
            *bundle
        )
    }

    private fun updatePlaybackState(@PlaybackStateCompat.State state: Int, vararg bundle: Bundle) {
        // 更新视频的总进度, setMetadata 会更新MediaControlCompat的onMetadataChanged
        val mPlaybackStateCompat = PlaybackStateCompat.Builder()
            .setState(
                state,
                exoPlayer.currentPosition,
                1.0f
            )
            .setActions(getAvailableActions(state))
            .setExtras(Arrays.asList(*bundle).getOrNull(0))
            .build()
        mediaSession.setPlaybackState(mPlaybackStateCompat)
    }

    override fun onPlay() {
        super.onPlay()
        if (playbackState.state != PlaybackStateCompat.STATE_PLAYING
            && myAudioManager.requestAudioFocus()
        ) {
            register()
            exoPlayer.play()
        }
    }

    fun sendMetadata() {
        val mediaItem = exoPlayer.currentMediaItem
        var mediaMetadataCompat: MediaMetadataCompat? = null
        if (mediaItem == null) {
            val currentSong = instance!!.currentSong
            if (currentSong != null) {
                mediaMetadataCompat =
                    currentSong.mapToMediaMetadata(exoPlayer.currentPosition, exoPlayer.duration)
            }
        } else {
            mediaMetadataCompat = getMediaMetadataCompat(mediaItem, exoPlayer.contentDuration)
        }
        if (mediaMetadataCompat != null) {
            mediaSession.setMetadata(mediaMetadataCompat)
        }
    }

    private val playbackState: PlaybackStateCompat
        private get() = mediaSession.controller.playbackState

    override fun onPause() {
        super.onPause()
        d(TAG, "onPause")
        if (playbackState.state == PlaybackStateCompat.STATE_PLAYING) {
            exoPlayer.pause()
            updatePlaybackState(PlaybackStateCompat.STATE_PAUSED)
            myAudioManager.unregisterBecomingNoisyReceiver()
            myAudioManager.abandonAudioFocus()
        }
    }

    override fun onStop() {
        super.onStop()
        exoPlayer.stop()
    }

    override fun onPlayFromUri(uri: Uri, extras: Bundle) {
        super.onPlayFromUri(uri, extras)
        val songs: List<Song>? = extras.getParcelableArrayList(Key.songList)
        if (songs != null) {
            val index = songs.indexOfFirst { song: Song -> song.url == uri.toString() }
            instance!!.playQueue = songs
            val mediaItems = songs.map { obj: Song -> obj.mapToExoPlayerMediaItem() }
            exoPlayer.setMediaItems(mediaItems, index, 0)
        } else {
            exoPlayer.setMediaItem(MediaItem.fromUri(uri))
        }
        updatePlaybackState(PlaybackStateCompat.STATE_CONNECTING)
        exoPlayer.prepare()
        onPlay()
    }

    override fun onSkipToNext() {
        super.onSkipToNext()
        exoPlayer.seekToNext()
        sendMetadata()
    }

    override fun onSkipToPrevious() {
        super.onSkipToPrevious()
        exoPlayer.seekToPrevious()
        sendMetadata()
    }

    override fun onSeekTo(pos: Long) {
        super.onSeekTo(pos)
        Log.d(TAG, "onSeekTo: $pos")
        exoPlayer.seekTo(pos)
        updatePlaybackState(PlaybackStateCompat.STATE_PLAYING)
    }

    override fun onSetPlaybackSpeed(speed: Float) {
        super.onSetPlaybackSpeed(speed)
    }

    override fun onSetRepeatMode(repeatMode: Int) {
        super.onSetRepeatMode(repeatMode)
        val playModel =
            if (repeatMode == PlaybackStateCompat.REPEAT_MODE_ONE) PlayMode.SINGLE_LOOP else PlayMode.ORDERLY
        setPlayMode(playModel)
        getInstance(service.application)!!.playMode = playModel
    }

    override fun onSetShuffleMode(shuffleMode: Int) {
        super.onSetShuffleMode(shuffleMode)
        setPlayMode(PlayMode.SHUFFLE)
        getInstance(service.application)!!.playMode = PlayMode.SHUFFLE
    }

    override fun unregister() {
        super.unregister()
        exoPlayer.release()
        mediaSession.release()
    }

    override fun onCustomAction(action: String, extras: Bundle) {
        super.onCustomAction(action, extras)
        when (action) {
            ACTION_SAVE_CURRENT_SONG_TO_FAVORITE -> {
                val mediaItem = exoPlayer.currentMediaItem
                if (mediaItem == null || mediaItem.localConfiguration == null) return
                val song =
                    instance!!.playQueue.firstOrNull { song1: Song -> song1.uniqueId == mediaItem.mediaId }
                instance!!.saveFavoriteSong(instance!!.currentSong)
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe({ song1: Song? -> toast(service, "收藏成功") }) { e: Throwable ->
                        e.printStackTrace()
                        toast(service, "收藏失败：" + e.message)
                    }
            }
        }
    }

    init {
        initPlayer(mediaSession)
    }

    companion object {
        private const val TAG = "MediaPlayerSession"
        fun getAvailableActions(@PlaybackStateCompat.State state: Int): Long {
            var actions = (PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS
                    or PlaybackStateCompat.ACTION_SKIP_TO_NEXT
                    or PlaybackStateCompat.ACTION_REWIND
                    or PlaybackStateCompat.ACTION_SEEK_TO)
            actions = if (state == PlaybackStateCompat.STATE_PLAYING) {
                actions or PlaybackStateCompat.ACTION_PAUSE
            } else {
                actions or PlaybackStateCompat.ACTION_PLAY
            }
            return actions
        }

        const val ACTION_SAVE_CURRENT_SONG_TO_FAVORITE = "action_save_to_favorite"
        fun switchPlayMode(
            transportControls: MediaControllerCompat.TransportControls,
            playMode: Int
        ) {
            if (playMode == PlayMode.ORDERLY) {
                transportControls.setRepeatMode(PlaybackStateCompat.REPEAT_MODE_ONE)
            } else if (playMode == PlayMode.SINGLE_LOOP) {
                transportControls.setShuffleMode(PlaybackStateCompat.SHUFFLE_MODE_ALL)
            } else {
                transportControls.setRepeatMode(PlaybackStateCompat.REPEAT_MODE_ALL)
            }
        }
    }
}